ruby/ql/lib/codeql/ruby/security/ImproperMemoizationQuery.qll

```diff
@@ -0,0 +1,87 @@
+/**
+ * Provides predicates for reasoning about improper memoization methods.
+ */
+
+private import ruby
+private import codeql.ruby.DataFlow
+private import codeql.ruby.dataflow.internal.DataFlowDispatch
+
+/**
+ * A `||=` statement that memoizes a value by storing it in an instance variable.
+ */
+private class MemoStmt extends AssignLogicalOrExpr {
+  private InstanceVariableAccess instanceVariable;
+
+  MemoStmt() { instanceVariable.getParent*() = this.getLeftOperand() }
+
+  /**
+   * Gets the variable access on the LHS of this statement.
+   * This is the `@a` in `@a ||= e`.
+   */
+  VariableAccess getVariableAccess() { result = instanceVariable }
+}
+
+/**
+ * A `||=` statement that stores a value at a particular location in a Hash,
+ * which itself is stored in an instance variable.
+ * For example:
+ * ```rb
+ * @a[key] ||= e
+ * ```
+ */
+private class HashMemoStmt extends MemoStmt {
+  HashMemoStmt() {
+    exists(ElementReference e |
+      e = this.getLeftOperand() and this.getVariableAccess().getParent+() = e
+    )
+  }
+}
+
+/**
+ * A method that may be performing memoization.
+ */
+private class MemoCandidate extends Method {
+  MemoCandidate() { exists(MemoStmt m | m.getEnclosingMethod() = this) }
+}
+
+/**
+ * Holds if parameter `p` of `m` is read in the right hand side of `assign`.
+ */
+private predicate parameterUsedInMemoValue(Method m, Parameter p, MemoStmt a) {
+  p = m.getAParameter() and
+  a.getParent+() = m and
+  p.getAVariable().getAnAccess().getParent*() = a.getRightOperand()
```

Comment:
I'm always a little suspicious of any use of `getParent()` in queries – ideally there should be some more specific way of navigating the AST like that. Would writing line 52 as `a.getEnclosingMethod() = m` mean the same thing?